### ISSUE TITLE
Inductor issue for CPU with int32 addition

### ISSUE THREAD

**Issue by davidradl** [pytorch/pytorch#108520]
### 🐛 Describe the bug

As a follow on from #108513 the following add operation gives a different output on torch and dynamo on CPU:

```
import torch
import torch._dynamo
import torch._dynamo.config
def mySum64(x):
    return (x+x).to(torch.int64)

x = torch.tensor( (2147483647), dtype=torch.int32)
torchResult = mySum64(x)
dynamoResult = torch.compile(mySum64)(x)

print(torchResult)
print(dynamoResult)
```

The output is :
```
tensor(-2)
tensor(4294967294)
```  
It looks like the compiler is not honouring the int32 for the add; instead it is incorrectly treating it as an int64. 

When we fix this a test should be added for this situation in the same place as #108513.

### Versions

Collecting environment information...
PyTorch version: 2.1.0a0+gite68b3ad
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.5.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:41:52) [Clang 15.0.7 ] (64-bit runtime)
Python platform: macOS-13.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==0.960
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] torch==2.1.0a0+git1004706
[pip3] torchgen==0.0.1
[conda] numpy                     1.25.2                   pypi_0    pypi
[conda] torch                     2.1.0a0+git1004706           dev_0    <develop>
[conda] torchgen                  0.0.1                    pypi_0    pypi

**Comment by lezcano** [comment]
This looks like an apple bug. I cannot repro on a linux box.

**Comment by peterbell10** [comment]
I can't reproduce either but it's probably just because clang tends to use undefined behavior to optimize more aggressively. Technically `4294967294` is a valid result.

**Comment by lezcano** [comment]
Right, overflow on signed types is UB, so I wouldn't worry too much about this case.

**Comment by jansel** [comment]
If that is true, then if you recompile PyTorch with clang the behavior of eager will change to match.

**Comment by peterbell10** [comment]
Eager won't ever give 4294967294 because we store `x + x`  in memory as a 32-bit integer, which forces the 64-bit extension to happen last. 

However, when you have the fused kernel `static_cast<long>(x + x)` the compiler can legally do the 64-bit extension first before the addition because it's equivalent in all cases that don't overflow.

**Comment by ezyang** [comment]
TBH, we probably should force signed overflow to be undefined, unlike in traditional C optimization we're not trying to optimize away bounds checks by reasoning about overflow.

**Comment by davidradl** [comment]
Are we aware of ML cases what would expect adding signed big numbers to turn negative after the overflow. I suspect this is likely to introduce unexpected behaviour in the model - without any warnings / errors - just negative numbers that might be made 0 by reLU or the like. Maybe for training a model we should have a flag to error for overflows, so the user knows that something potentially unexpected is happening and can then intentionally mitigate for those data values. I agree with @ezyang; it seems like a way we can standardise behaviour across the kernels, reducing kernel specific differences.

**Comment by ezyang** [comment]
Well, we're specifically talking about int tensors, so unlikely to show up in relu. However, with int8, overflow is kind of... a fact of life? @ar0ck do you mind telling us a little more about what was going on when you ran into this problem at https://github.com/pytorch/pytorch/issues/105264 ?

**Comment by ar0ck** [comment]
@ezyang, this occurred in the context of using PyTorch to simulate low-level HW operations for an accelerator. Think extracting bits, doing 8-bit math, upcasting, doing more math, etc. When attempting to speed things up via `torch.compile`, I ran into this discrepancy.

I think I agree with your comment - if eager mode defines signed integer overflow to be two's complement (does it? I can't find a reference), then compiled code ought to respect that.

**Comment by ezyang** [comment]
We don't define it per se, but because we don't fuse the integer operations per se, de facto we will do signed overflow because there's no opportunity for the compiler to change it up.

**Comment by leslie-fang-intel** [comment]
Undefined behavior between different compiler is expected. Feel free to re-open if any further discussion needed here.
